File: cassandra-service-base/src/main/scala/com/github/sedovalx/cassandra/services/base/AbstractAccessorJava8Adapter.scala
package com.github.sedovalx.cassandra.services.base

import java.util._
import java.util.concurrent.CompletableFuture

import com.datastax.driver.core.{ResultSet, ResultSetFuture}
import com.datastax.driver.mapping.Result
import com.github.sedovalx.cassandra.services.utils.Java8Interop
import com.github.sedovalx.cassandra.services.utils.Java8Interop._
import com.google.common.util.concurrent.ListenableFuture
import net.javacrumbs.futureconverter.java8guava.FutureConverter

/**
  * Author alsedov on 29.12.2015
  */
abstract class AbstractAccessorJava8Adapter[Entity] {
    protected def toCompletableFutureEntity(future: ListenableFuture[Entity]): CompletableFuture[Optional[Entity]] = {
        FutureConverter.toCompletableFuture(future).thenApply(toJavaFunction(toOptional))
    }

    protected def toCompletableFutureResult(future: ListenableFuture[Result[Entity]]): CompletableFuture[Result[Entity]] = {
        FutureConverter.toCompletableFuture(future)
    }

    protected def toCompletableFutureResultSet(future: ResultSetFuture): CompletableFuture[ResultSet] = {
        FutureConverter.toCompletableFuture(future)
    }

    protected def toVoidFutureResultSet(future: ResultSetFuture): CompletableFuture[Void] = {
        toCompletableFutureResultSet(future).thenApply { toJavaFunction(_ => null) }
    }

    protected def toVoidFuture(future: ListenableFuture[Result[Entity]]): CompletableFuture[Void] = {
        FutureConverter.toCompletableFuture(future).thenApply { toJavaFunction(_ => null) }
    }

    protected def toOptional(entity: Entity): Optional[Entity] = {
        Java8Interop.toOptional(entity)
    }
}

File: cassandra-service-base/src/main/java/com/github/sedovalx/cassandra/services/base/AbstractAccessorJava8Adapter.java
package com.github.sedovalx.cassandra.services.base;

import com.datastax.driver.core.ResultSet;
import com.datastax.driver.core.ResultSetFuture;
import com.datastax.driver.mapping.Result;
import com.google.common.util.concurrent.ListenableFuture;
import net.javacrumbs.futureconverter.java8guava.FutureConverter;

import java.util.Optional;
import java.util.concurrent.CompletableFuture;

/**
 * Created by Alexander
 * on 22.03.2016.
 */
public abstract class AbstractAccessorJava8Adapter<Entity> {
    protected CompletableFuture<Optional<Entity>> toCompletableFutureEntity(ListenableFuture<Entity> future) {
        return FutureConverter.toCompletableFuture(future).thenApply(Optional::ofNullable);
    }

    protected CompletableFuture<Result<Entity>> toCompletableFutureResult(ListenableFuture<Result<Entity>> future) {
        return FutureConverter.toCompletableFuture(future);
    }

    protected CompletableFuture<ResultSet> toCompletableFutureResultSet(ResultSetFuture future) {
        return FutureConverter.toCompletableFuture(future);
    }

    protected CompletableFuture<Void> toVoidFutureResultSet(ResultSetFuture future) {
        return FutureConverter.toCompletableFuture(future).thenApply(rows -> null);
    }

    protected CompletableFuture<Void> toVoidFuture(ListenableFuture<Result<Entity>> future) {
        return FutureConverter.toCompletableFuture(future).thenApply(entities -> null);
    }

    protected Optional<Entity> toOptional(Entity entity) {
        return Optional.ofNullable(entity);
    }
}
